docs/decisions/0021-json-serializable-custom-types.md

```diff
@@ -0,0 +1,163 @@
+        if (targetType == typeof(long)) { return new JsonConverterWrapper(new Int64Converter()); }
+        if (targetType == typeof(float)) { return new JsonConverterWrapper(new SingleConverter()); }
+        if (targetType == typeof(double)) { return new JsonConverterWrapper(new DoubleConverter()); }
+        if (targetType == typeof(decimal)) { return new JsonConverterWrapper(new DecimalConverter()); }
+        if (targetType == typeof(TimeSpan)) { return new JsonConverterWrapper(new TimeSpanConverter()); }
+        if (targetType == typeof(DateTime)) { return new JsonConverterWrapper(new DateTimeConverter()); }
+        if (targetType == typeof(DateTimeOffset)) { return new JsonConverterWrapper(new DateTimeOffsetConverter()); }
+        if (targetType == typeof(Uri)) { return new JsonConverterWrapper(new UriTypeConverter()); }
+        if (targetType == typeof(Guid)) { return new JsonConverterWrapper(new GuidConverter()); }
+
+        if (targetType.GetCustomAttribute<JsonConverterAttribute>() is JsonConverterAttribute tca &&
+            Type.GetType(tca.ConverterType.Name, throwOnError: false) is Type converterType &&
+            Activator.CreateInstance(converterType) is JsonConverter converter)
+        {
+            return converter;
+        }
+
+        return new SimpleJsonConverter();
+    }
+
+    private class JsonConverterWrapper : JsonConverter<object> {
+        private readonly TypeConverter _converter;
+
+        public JsonConverterWrapper(TypeConverter converter)
+        {
+            this._converter = converter;
+        }
+
+        public override object Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            return this._converter.ConvertFromString(null, CultureInfo.InvariantCulture, reader.GetString());
+        }
+
+        public override void Write(Utf8JsonWriter writer, object value, JsonSerializerOptions options)
+        {
+            writer.WriteStringValue(this._converter.ConvertToString(null, CultureInfo.InvariantCulture, value));
+        }
+    }
+
+
+    private class SimpleJsonConverter : JsonConverter<object>
+    {
+        public override object? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            return JsonSerializer.Deserialize(ref reader, typeToConvert, options);
+        }
+
+        public override void Write(Utf8JsonWriter writer, object value, JsonSerializerOptions options)
+        {
+            JsonSerializer.Serialize(writer, value, value.GetType(), options);
+        }
+    }
+
+```
+
+If a developer wants to customize the serialization and/or deserialization of a custom type, they can implement their own `JsonConverter` and apply the `JsonConverterAttribute` their custom type as seen below.
+
+**Example usage of the `JsonConverter`** class and attribute:
+
+```csharp
+    [JsonConverter(typeof(MyCustomTypeConverter))]
+    private sealed class MyCustomType
```

Comment:
The example above uses a `TypeConverter` which is implemented using JSON so it does two/from strings but it is flexible enough to support any type. Are we changing that?

Reply:
updated to use a `TypeConverter` but then return a default, JSON-serializing one if no custom converter is found